libethashseal/Ethash.h

```diff
@@ -92,7 +121,16 @@ class Ethash: public SealEngineBase
 
     /// A mutex covering m_sealing
     Mutex m_submitLock;
+
+    u256 externalHashrate() const;
+
+    // external hashrate
+    mutable std::unordered_map<h256, std::pair<u256, std::chrono::steady_clock::time_point>>
+        m_externalRates;
+    mutable SharedMutex x_externalRates;
 };
 
+DEV_SIMPLE_EXCEPTION(InvalidSealEngine);
+Ethash& asEthash(SealEngineFace* _f);
```

Comment:
Maybe change this to get reference parameter, it's more logical to pass reference and transform it to reference,